[4단계 - Transaction synchronization 적용하기] 줄리(한승아) 미션 제출합니다.

jbilee opened the pull request:
안녕하세요, 모다!! 🤗

마지막 단계 PR 올립니다! 이전 단계에서 DAO를 통해 Connection을 전달하던 로직을 DataSourceUtils와 TransactionSynchronizationManager로 대체했습니다.

3단계를 진행할 때까지만 해도 DataSource로 Connection을 직접 제어하는 로직이 서비스에 구현돼있다는 점이 어색하게 느껴졌었는데, 다시 생각해 보니 비즈니스 로직에서의 트랜잭션 경계는 서비스만이 알 수 있는 부분인 것 같습니다. 하여 서비스가 Connection을 제어하는 구조를 유지했습니다.

### 이번 단계 고민 포인트
- Auto-commit의 on/off 여부에 따라 Connection을 닫는 로직을 서비스 또는 JdbcTemplate에 두었는데, 어색한 점이 있을지 혹은 제가 놓친 부분이 있을지 궁금합니다!

남은 리뷰도 잘 부탁드립니다, 감사합니다~! 🙏

jbilee committed:
feat: DataSourceUtils 및 TransactionSynchronizationManager 내부 기능 구현

jbilee committed:
refactor: ConnectionManager를 TransactionManager로 교체

jbilee committed:
test: 코드 변경으로 인한 테스트 코드 및 mock 업데이트

jbilee committed:
fix: 트랜젝션 제어 관련 오류 수정

- 단발성 JdbcTemplate 메서드 호출에서는 커넥션을 닫지 않던 이슈 해결
- 트랜잭션 경계가 예상한 대로 생성되도록 수정

jbilee committed:
refactor: JdbcTemplate 내에서 커넥션을 닫는 로직을 finally 블록으로 이동

jbilee committed:
refactor: 서비스 로직 내 예외 처리 개선

jbilee commented on jdbc/src/main/java/com/interface21/jdbc/core/JdbcTemplate.java:
```diff
@@ -69,11 +53,18 @@ public <T> List<T> queryMany(final String sql, final ResultSetMapper<T> resultSe
             setParameters(pstmt, parameters);
-            return queryExecutor.run(pstmt);
+            T result = queryExecutor.run(pstmt);
+
+            if (connection.getAutoCommit()) {
+                DataSourceUtils.releaseConnection(connection, dataSource);
+            }
+
+            return result;
         } catch (SQLException e) {
```
[refactor: JdbcTemplate 내에서 커넥션을 닫는 로직을 finally 블록으로 이동](https://github.com/woowacourse/java-jdbc/pull/1136/commits/9f4596463831888475be7a5f3a94089f06c14578)

jbilee commented on app/src/main/java/com/techcourse/service/TxUserService.java:
```diff
@@ -0,0 +1,56 @@
+                connection.rollback();
+            } catch (SQLException e2) {
+                throw new CannotGetJdbcConnectionException("Failed to rollback");
+            }
```
[refactor: 서비스 로직 내 예외 처리 개선](https://github.com/woowacourse/java-jdbc/pull/1136/commits/8b40bf364e07aad4f9dd59eceeaeedf2f1e8400d)

Chaeyoung714 commented on jdbc/src/main/java/com/interface21/jdbc/datasource/DataSourceUtils.java:
```diff
@@ -29,6 +30,9 @@ public static Connection getConnection(DataSource dataSource) throws CannotGetJd
         try {
+            if (!connection.getAutoCommit()) {
+                TransactionSynchronizationManager.unbindResource(dataSource);
+            }
```
이러한 분기문을 둔 이유가 궁금해요! autoCommit = true라면 unbindResource를 해도 null로 반환되니깐 분기문이 없어도 되지 않을까요?

jbilee replied:
동의합니다! Bind된 리소스가 없다 하더라도 예외가 발생할 일이 없어서, if문을 제거하였습니다! 👍

[refactor: DataSourceUtils 내 불필요한 if 검증 제거](https://github.com/woowacourse/java-jdbc/pull/1136/commits/12a1b3fae04d489e67667b3c55744a8ddf157dbb)

jbilee committed:
fix: 커넥션 close 관련 논리적 오류 수정

- JdbcTemplate에서 auto-commit 옵션이 true일 때만 자체적으로 커넥션을 close하도록 수정

Chaeyoung714 commented on app/src/main/java/com/techcourse/service/TxUserService.java:
```diff
@@ -0,0 +1,62 @@
+        this.dataSource = DataSourceConfig.getInstance();
+    }
+
+    @Override
+    public User findById(final long id) {
+        return userService.findById(id);
+    }
+
+    @Override
+    public User findByAccount(final String account) {
+        return userService.findByAccount(account);
+    }
+
+    @Override
+    public void save(final User user) {
+        userService.save(user);
+    }
```
findById, findByAccount, save 메서드는 트랜잭션 경계 설정을 하지 않은 이유가 궁금해요!
위 메서드들은 UserService 단에서 트랜잭션 경계 설정을 하지 않았기 때문에 줄리가 releaseConnection을 autoCommit 여부에 따라 분기처리해주었다고 이해했는데요, 모든 UserService에 트랜잭션 경계를 추가한다면 autoCommit = false인 커밋만 하게 되니까 분기처리가 필요없어진다고 생각했어요!

위의 메서드들은 Sql을 단건만 보낸다는 특성이 있는데, 트랜잭션 경계를 설정하지 않고 autoCommit = true로 보내는 이유가 있을까요?
TxUserService 입장에서는 UserService(appUserService)가 Sql문을 단건만 보내는지 여러건을 보내는지 알고 경계를 설정할 필요가 있을까요? 줄리의 생각이 궁금합니다!

jbilee replied:
모다가 질문해주신 것처럼 TxUserService 입장에서 AppUserService가 SQL문을 단건 혹은 여러 건으로 보내는지는 알 필요가 없다고 생각합니다! 다만 현재 구조에서는 아직 AppUserService도 커넥션을 다룰 수 있기 때문에, 혹여나 TxUserService가 아닌 AppUserService의 메서드를 호출하는 경우에 대비하여 JdbcTemplate 단에서도 커넥션을 닫을 수 있어야 한다고 판단했습니다.

하여 auto-commit 상태를 확인하는 분기문은 유지하지만, TxUserService의 다른 메서드들도 직접적으로 트랜잭션 경계를 제어하는 방향으로 수정해보았습니다! 👀

[refactor: TxUserService 내 모든 메서드에 트랜잭션 경계 적용](https://github.com/woowacourse/java-jdbc/pull/1136/commits/91214c492b2bcac6b561c71abdcac46801be671e)

Chaeyoung714 replied:
아하~ 이해됐어요! 이렇게 하면 캡슐화는 더 잘 되면서도 확장성은 유지할 수 있겠네용! 좋습니다!

Chaeyoung714 reviewed (changes requested):
줄리~~ 모다입니다! 이번 미션도 고생많으셨어요!

테스트 관련 피드백도 바로 확인하고 반영해주셔서 감사합니다!
줄리가 말한 autoCommit 관련 부분도 확인했고, 이것과 관련해서 궁금한 점을 코멘트로 달아두었어요!
요구사항은 충분히 만족했지만, 마지막 미션인만큼 rc로 남겨두겠습니다!

그럼 확인 부탁드려용 (___)꾸벅

jbilee committed:
refactor: DataSourceUtils 내 불필요한 if 검증 제거

- 트랜잭션 매니저에서 리소스를 unbind하는 로직은 bind된 리소스가 없어도 예외가 발생하지 않기 때문에 auto-commit 여부와 상관없이 호출하도록 하여 if문 제거

jbilee committed:
refactor: TxUserService 내 모든 메서드에 트랜잭션 경계 적용

Chaeyoung714 reviewed (approved):
줄리~~ 코멘트랑 변경사항 잘 확인했어용! 
이번 미션 여기서 머지하겠습니다!
레벨4 정말 고생 많으셨고 마지막 리뷰어였어서 영광이었어요 😊 앞으로도 화이팅!!
